### PR TITLE
Reduce API calls from certain elements by keeping global state up-to-date and caching redundant data

### PR DESCRIPTION
**Pull request by pinheadmz**
Closes https://github.com/kyokan/bob-wallet/issues/100

This PR reduces API calls in certain views to improve UI performance. The "Your Bids" view in particular has a table with three elements per bid: each of these elements was individually calling `getNameInfo()` on the same name! In addition, the use of `componentWillMount()` in certain elements to update state is insufficient since a new block event (and state update) will not trigger this method. The elements themselves should just rely on the global state being updated already.

Since bids and namestates only change when new blocks connect to the chain, we only really need to gather this data once per block. The data is already stored in the global state, so all we have to do is change _when_ it is updated.

This patch adds a new function `onNewBlock()` to `backgroundMonitor` which is executed any time a `doPoll()` encounters a new block from hsd. This function gets current fee rates, gets all bids, and then iterates through all the bids getting updated namestates. This means that on every new block, UI may be a bit sluggish for a minute. A new progress indicator is added in the sidebar footer.

Tested by running a bash script that opened 300 names and then bid once or twice on each name.

Future work:
- Add pagination to hsd wallet (see https://github.com/bcoin-org/bcoin/pull/605)
- Add functionality either in Bob or hsd to filter out expired auctions or bids that "don't matter anymore" for one reason or another.
- Switch the hsd<->Bob link from 10-second polling to **websocket events**. If possible, this will keep Bob up to date with essentially zero lag.
- Explore `getNameInfo` for ways to reduce 2 API calls (`node.getNameInfo` + `wallet.getAuctionInfo`) to one, or cache some data if possible.

### Demonstration

UI elements seem more responsive than before even when a new block is being processed. Once the new block processing is done, everything is perfectly snappy since no new API calls are made when loading/changing views.

Note the new "New Block Status" message just above the current height in the sidebar footer.

LoadBidsOnNewBlock

### CODE REVIEW

**Comment by pinheadmz** [comment]
rebase to 525a7e56f5a6d43a356d950b6b2e199bd5458758:
rebasing on master

**Comment by mslipper** [comment]
https://www.youtube.com/watch?v=VI6dsMeABpU. Will review in the AM.

**Comment by pinheadmz** [comment]
TODO: Use newBlockStatus to monitor the flurry of `getNameByHash` calls on boot

**Comment by mslipper** [comment]
LGTM. Let me know when you want a merge.

**Comment by pinheadmz** [comment]
@mslipper added one last patchset. This replaces the `state.wallet.transactions` array with a Map. The Map is still sorted by date, and gives us the feature of being able to test it for a certain txid. With that in place, another round of redundant API calls can be avoided.

Pushing this up for now, going to bang on it a bit more tomorrow and make sure it's all good.

File: app/ducks/walletReducer.js [hunk 1]
```diff
@@ -56,7 +56,7 @@ export default function walletReducer(state = getInitialState(), {type, payload}
           ...state.balance
         },
         isLocked: true,
-        transactions: []
+        transactions: new Map()
```
Comment (mslipper) [inline app/ducks/walletReducer.js:59]: Curious: what is the advantage of using a `Map` in this context as opposed to a simple `{}`?

Comment (pinheadmz) [reply app/ducks/walletReducer.js:59]: The PropTypes checker was throwing an error before the first update.

**Comment by pinheadmz** [comment]
rebase to 216a858cff00d5194c9109a37a89effa2f85a836

Remove leftover `console.log` from debugging and force-push

**Pull request merged**